xla/service/gpu/cudnn_fused_mha_rewriter.cc

```diff
@@ -337,150 +324,128 @@ std::vector<int64_t> GetDimensionVector(absl::Span<const int64_t> dimensions,
+
+  std::vector<int64_t> bmm2_bh =
+      GetDimensionVector(bmm_2->operand(0)->shape().dimensions(),
+                         bmm2_dnums.lhs_batch_dimensions());
+
+  std::vector<int64_t> bmm2_s_kv =
+      GetDimensionVector(bmm_2->operand(0)->shape().dimensions(),
+                         bmm2_dnums.lhs_contracting_dimensions());
+
+  std::vector<int64_t> bmm2_s_q =
+      need_canonicalization
+          ? GetDimensionVector(bmm_2->operand(1)->shape().dimensions(),
+                               bmm2_rhs_non_contracting_dims)
+          : GetDimensionVector(bmm_2->operand(0)->shape().dimensions(),
+                               bmm2_lhs_non_contracting_dims);
+
+  std::vector<int64_t> bmm2_d =
+      need_canonicalization
+          ? GetDimensionVector(bmm_2->operand(0)->shape().dimensions(),
+                               bmm2_lhs_non_contracting_dims)
+          : GetDimensionVector(bmm_2->operand(1)->shape().dimensions(),
+                               bmm2_rhs_non_contracting_dims);
+
+  TF_RET_CHECK(bmm2_bh.size() == 2);
+  TF_RET_CHECK(bmm2_s_q.size() == 1);
+  TF_RET_CHECK(bmm2_s_kv.size() == 1);
+  TF_RET_CHECK(bmm2_d.size() == 1);
+
+  // check if bhsd is correct between bmm1 and bmm2
+  if (bmm1_bh[0] != bmm2_bh[0] || bmm1_bh[1] != bmm2_bh[1] ||
+      bmm1_s_q[0] != bmm2_s_q[0] || bmm1_s_kv[0] != bmm2_s_kv[0] ||
+      bmm1_d[0] != bmm2_d[0]) {
+    return std::vector<int64_t>{};
+  }
+
+  std::vector<int64_t> bhsd = {bmm1_bh[0], bmm1_bh[1], bmm1_s_q[0],
+                               bmm1_s_kv[0], bmm1_d[0]};
+  return bhsd;
 }
 
-absl::StatusOr<bool> IsFlashAttention(
-    HloInstruction* bmm_1, bool is_causal_mask,
-    absl::string_view custom_call_name,
+absl::StatusOr<bool> IsFusedAttention(
+    std::vector<int64_t>& bhsd, bool is_training,
     stream_executor::CudaComputeCapability cc,
     stream_executor::dnn::VersionInfo cudnn_version) {
-  const DotDimensionNumbers& dnums = bmm_1->dot_dimension_numbers();
-  TF_ASSIGN_OR_RETURN(
-      std::vector<int64_t> seq_q_dims,
-      GetNonContractingDims(bmm_1->operand(0)->shape(),
-                            dnums.lhs_batch_dimensions(),
-                            dnums.lhs_contracting_dimensions()));
-
-  TF_ASSIGN_OR_RETURN(
-      std::vector<int64_t> seq_k_dims,
-      GetNonContractingDims(bmm_1->operand(1)->shape(),
-                            dnums.rhs_batch_dimensions(),
-                            dnums.rhs_contracting_dimensions()));
-
-  std::vector<int64_t> seq_q =
-      GetDimensionVector(bmm_1->operand(0)->shape().dimensions(), seq_q_dims);
-
-  std::vector<int64_t> seq_k =
-      GetDimensionVector(bmm_1->operand(1)->shape().dimensions(), seq_k_dims);
+  // otherwise check if it is supported by regular attention
+  int64_t s_q = bhsd[2];
+  int64_t s_kv = bhsd[3];
+  int64_t hidden_dim = bhsd[4];
+  bool is_seqlen_supported = (s_q <= 512 && s_kv <= 512) &&
+                             (!is_training || s_q % 64 == 0 && s_kv % 64 == 0);
```

Comment:
Please put () around the extra && condition. This would trigger a compiler warning otherwise which we treat internally as error.